### ISSUE TITLE
DELEGATECALL vs CALL vs CALLCODE for second layer solutions

### ISSUE THREAD

**Issue by tpmccallum** [ethereum/vyper#960]
### What's your issue about?

Please include information like:

* This is a question

Could someone please let me know if Vyper will allow the use of DELEGATECALL OPCODE in its smart contracts? 
I see that the DELEGATECALL OPCODE is in the opcode file [1]

* Here is some context

As I understand it, Vyper uses the CALL OPCODE for safety/security reasons. This makes sense. However, there is currently an immense amount of designing and coding of second layer solutions (side chains etc.). These designs and implementations require interactive games (whereby participants on the side chain can provide proofs and so forth). These interactive games require smart contracts to perform DELEGATECALLs as apposed to just CALLs.

### How can it be fixed?

I would like to include this information in the Vyper chapter of Mastering Ethereum. I will be happy to just report the definitive answer. There is no fix required, it just is what it is. I think it would be really useful for developers to understand this important difference between Vyper and say Solidity in the event that they are designing and building second layer solutions.

![image](https://user-images.githubusercontent.com/9831342/43041326-5889425a-8da0-11e8-9a2c-f4752a07bc31.png)

Many thanks
Tim

[1] https://github.com/ethereum/vyper/blob/master/vyper/opcodes.py

**Comment by fubuloubu** [comment]
I agree on the library insofar that it would be the first use case in Vyper that would directly require it (internal calls can use it as an optimization strategy, but I think that's out of scope for this discussion).

Supporting on-chain libraries would be a major feature to add to Vyper. It is not currently supported. We would have to discuss a mechanism for making sure security is respected when adding such a feature, and I have a feeling this is a longer term discussion involving ethPM integration and some checks of proper-initialization of the library. Uninitialized or improperly utilized libraries are a major source of security issues for the "not well versed" in EVM fundamentals, so a properly implemented feature with the Vyper Goal of making it maximally difficult to write insecure code would probably look much different than the equivalent Solidity.

---

Syntax-wise @jacqueswww I think that might work, but as I note above a LOT more considerations will be need to taken in account for how this feature looks in practice. I would say it's out of scope for Vyper's first release.
